src/scene/materials/shader-graph-node.js

```diff
@@ -0,0 +1,615 @@
+        for (i = 0; i < this.graphData.ioPorts.length; i++) {
+            var matId = '_' + this.id;
+            ioPort = this.graphData.ioPorts[i];
+            if (ioPort.name.startsWith('IN_') || (ioPort.name.startsWith('CONST_') && ioPort.type === 'sampler2D')) {
+                generatedGlsl += 'uniform ' + ioPort.type + ' ' + ioPort.name + matId + ';\n';
+            }
+        }
+        // run through constants values are set here (except for textures - which have to be uniforms)
+        for (i = 0; i < this.graphData.ioPorts.length; i++) {
+            ioPort = this.graphData.ioPorts[i];
+            if (ioPort.name.startsWith('CONST_') && (ioPort.type != 'sampler2D' )) {
+                generatedGlsl += ioPort.type + ' ' + ioPort.name + ' = ' + this._getIoPortValueString(ioPort) + ';\n';
+            }
+        }
+
+        // get all sub graph function definitions (including functions in sub graphs' sub graphs ...)
+        // assumes names are unique - maybe should be id or key?
+        var depGraphFuncs = {};
+        var depIoPortList = [];
+
+        var depName = this.name;
+        if (!this.graphData.customFuncGlsl) {
+            depName += '_' + this.id;
+        }
+        depGraphFuncs[depName] = this._generateFuncGlsl(); // this should prevent infinite recursion?
+
+        this._generateSubGraphFuncs(depGraphFuncs, depIoPortList);
+
+        // declare all dependancy textures
+        for (i = 0; i < depIoPortList.length; i++) {
+            generatedGlsl += depIoPortList[i];// .declString;
+        }
+
+        // add all the graph definitions
+        var depGraphList = []; // reverse order
+        for (var func in depGraphFuncs) {
+            var funcString = '';
+
+            if ( func.endsWith('PS') ) {
+                funcString += '#ifdef SG_PS\n';
+            } else if ( func.endsWith('VS') ) {
+                funcString += '#ifdef SG_VS\n';
+            }
+
+            funcString += depGraphFuncs[func] + '\n';
+
+            if ( func.endsWith('PS')  ) {
+                funcString += '#endif //SG_PS\n';
+            } else if ( func.endsWith('VS') ) {
+                funcString += '#endif //SG_VS\n';
+            }
+
+            depGraphList.push(funcString);
+        }
+        var tLen = depGraphList.length; // need this because pop() reduces array length!
+        for (i = 0; i < tLen; i++) {
+            generatedGlsl += depGraphList.pop();
+        }
+
+        return generatedGlsl;
+    },
+
+    generateRootCallGlsl: function () {
+        var generatedGlsl = '';
+
+        // generate input and output names for function call and run through outputs to declare variables
+        var inNames = {};
+        var outNames = {};
+
+        for (var i = 0; i < this.graphData.ioPorts.length; i++) {
+            var matId = '_' + this.id;
+            var ioPort = this.graphData.ioPorts[i];
+            if (ioPort.name.startsWith('IN_')) {

```

Comment:
Personally I'd be tempted to make Port a class, and when its constructed with name, I would store booleans on it:
```
this.isInput = this.name.startsWith('IN_');
this.isOutput = this.name.startsWith('OUT_');
...
```

Reply:
I am considering refactoring to split out ports into categories - or maintain a transient private structure to accelerate access and/or group the split categories for convenient iteration.

Reply:
yep either of these would work well

Reply:
I have refactored and removed prefixes entirely and replaced with PORT_TYPE_ enum - overall it is improved I think - thanks!